R/submit_ipr.R

```diff
@@ -1,145 +1,417 @@
+    idx <- 1L
+    batch_size_default <- 30L
+    n_submitted <- 0L
+    job_ids <- c() # job id vector
+    while ((n_seqs > 0L) || ((.get_running_jobs()) > 0L)) {
+      # only allow 30 job submissions at a time
+      n_available <- 30L - .get_running_jobs() # .get_running_jobs() return N running jobs
+      n_submissions <- min(n_seqs, batch_size_default, n_available) # minimum of available seqs for submission
+
+      if (n_submissions > 0L) {
+        # stop_idx has 2 conditions (1st batch submit && 2..N batch submit)
+        stop_idx <- ifelse(n_submitted < 30, n_submissions, (idx + n_submissions - 1))
+
+        cat("Submitting a batch of", (stop_idx - idx), "seqs\n")
+        for (i in idx:stop_idx) {
+          fasta <- file.path(split_seqs_folder, fasta_files[i])
+          cat("Submitting the ", i, "th", "sequence: ", fasta, " for analysis\n", sep = "")
+          job_id <- .submit(fasta, user_email)
+          cat("Returned job id: ", job_id, "\n")
+          if (job_id != "failed") { # .submit() will return either job_id or 'failed'
+            job_ids <- append(job_ids, job_id)
+          }
+        }
+        print("job_ids vec: ")
+        print(job_ids)
+        idx <- idx + n_submissions
+        n_seqs <- n_seqs - n_submissions
+        n_submitted <- n_submitted + n_submissions
+      }
 
-  cat("Multiple sequences detected", "\n", sep = "")
-  split_seqs_folder <- .split_seqs(path2seq, outfolder)
-
-  seq_i <- 1L
-  seq_n <- length(list.files(split_seqs_folder))
-  n_successes <- 0L
-  for (seq_cur in list.files(split_seqs_folder)) {
-    seq_cur_path <- file.path(split_seqs_folder, seq_cur)
-    outfile <- file.path(outfolder, strsplit(seq_cur, ".faa"[[1]]))
-    success <- as.integer(.submit(seq_cur_path, outfile, email))
-    cat("Seq # ", seq_i, "/", seq_n, " completed\n", sep = "")
-    seq_i <- seq_i + 1L
+      # poll each job & update status table
+      for (job in job_ids) {
+        status <- .poll_job(job)
+        .update_status_table(job, status)
+        # write/get result (if completed)
+        if (status == "completed") {
+          .get_job_result(job, outfolder, write_result = TRUE)
+          .update_status_table(job, status)
+          job_ids <- job_ids[job_ids != job] # rm job id from vec when completed
+        } else {
+          n_polls <- .lookup_polls(job)
+          cat("tb_status:\n")
+          print(tb_status)
+          cat("N polls for ", job, ": ", n_polls, "\n", sep = "")
+          # forget jobs that are polled more than N times
+          if (n_polls > 500) {
+            job_ids <- job_ids[job_ids != job]
+            .update_status_table(job_code = job, status_update = "failed")
+          }
+        }
+      }
+      cat("Number of sequences remaining: ", n_seqs, "\n", sep = "")
+      cat("N running jobs: ", .get_running_jobs(), "\n", sep = "")
+      cat("Total jobs: ", total_jobs, "\n", sep = "")
+      Sys.sleep(5L)
+    }
   }
+  ### END job processing
```

Comment:
So, feel free to address this or not, since this is a demo package for a workshop, but I didn't find it obvious that the single-sequence case never reaches this point, since it returns in all of the possible paths in its case.

I can think of a number of ways to fix it: the least troublesome would be to pull out the single and multi-sequence processing into separate functions, which would make the logic in this function would be more apparent. I'd put the parts that follow `### END job processing` just in the multi-sequence function, since they're only used in that case. Alternatively, you could not treat the single-sequence case as a special case, and instead just treat it as a multi-sequence case where the number of sequences happens to be one.

Reply:
I like the multi being the only/default option, where n=1 in `single`. Jake, any downsides that you note for this?